lib/EarlyFunctionAnalysis/EarlyFunctionAnalysis.cpp

```diff
@@ -843,6 +878,98 @@ void FunctionEntrypointAnalyzer::interproceduralPropagation() {
   }
 }
 
+std::optional<abi::RegisterState::Values>
+FEA::tryGetRegisterState(model::Register::Values RegisterValue,
+                         const ABIAnalyses::RegisterStateMap &ABIRegisterMap) {
+  using State = abi::RegisterState::Values;
+
+  llvm::StringRef Name = model::Register::getCSVName(RegisterValue);
+  if (llvm::GlobalVariable *CSV = M.getGlobalVariable(Name, true)) {
+    auto It = ABIRegisterMap.find(CSV);
+    if (It != ABIRegisterMap.end()) {
+      revng_assert(It->second != State::Count && It->second != State::Invalid);
+      return It->second;
+    }
+  }
+
+  return std::nullopt;
+}
+
+void FEA::initializeMapForDeductions(FunctionSummary &Summary,
+                                     abi::RegisterState::Map &Map) {
+  auto Arch = model::ABI::getArchitecture(Binary->DefaultABI);
+  revng_assert(Arch == Binary->Architecture);
+
+  for (const auto &Reg : model::Architecture::registers(Arch)) {
+    const auto &ArgRegisters = Summary.ABIResults.ArgumentsRegisters;
+    const auto &RVRegisters = Summary.ABIResults.FinalReturnValuesRegisters;
+
+    if (auto MaybeState = tryGetRegisterState(Reg, ArgRegisters))
+      Map[Reg].IsUsedForPassingArguments = *MaybeState;
+
+    if (auto MaybeState = tryGetRegisterState(Reg, RVRegisters))
+      Map[Reg].IsUsedForReturningValues = *MaybeState;
+  }
+}
+
+void FunctionEntrypointAnalyzer::applyABIDeductions() {
+  using namespace abi;
+  using State = RegisterState::Values;
+
+  if (ABIEnforcement.getNumOccurrences() == 0
+      || ABIEnforcement == NoABIEnforcement)
+    return;
+
+  for (const model::Function &Function : Binary->Functions) {
+    auto &Summary = Oracle.at(Function.Entry);
+
+    RegisterState::Map StateMap(Binary->Architecture);
+    initializeMapForDeductions(Summary, StateMap);
+
+    bool EnforceABIConformance = ABIEnforcement == FullABIEnforcement;
+    std::optional<abi::RegisterState::Map>
+      ResultMap = EnforceABIConformance ?
+                    enforceRegisterStateDeductions(StateMap,
+                                                   Binary->DefaultABI) :
+                    tryApplyRegisterStateDeductions(StateMap,
+                                                    Binary->DefaultABI);
```

Comment:
```c++
    std::optional<abi::RegisterState::Map> ResultMap;
    if (EnforceABIConformance) {
      ResultMap = enforceRegisterStateDeductions(StateMap,
                                                   Binary->DefaultABI) 
    } else {
      ResultMap = tryApplyRegisterStateDeductions(StateMap,
                                                    Binary->DefaultABI);
    }
```

Reply:
This is not possible. `ResultMap` has a `const` field. And @antoniofrighetto predicted that you're going to hate the ternary operator :D

Reply:
A separate function like this might be a good compromise thought:
```cpp
std::optional<abi::RegisterState::Map> doEnforceSomethingOrOther(const abi::RegisterState::Map &StateMap, model::ABI::Values ABI) {
  if (EnforceABIConformance) {
    return enforceRegisterStateDeductions(StateMap,
                                               Binary->DefaultABI) 
  } else {
    return tryApplyRegisterStateDeductions(StateMap,
                                                Binary->DefaultABI);
  }
}
```

Reply:
Achieving this would require extra change to `abi::RegisterState::Map`: the copy assignment operator is implicitly deleted because that class has a private `const` field (expression not assignable). So, I eventually decided, for the sake of simplicity, to use the ternary operator. Does it sound OK to you?

Reply:
I think we should drop the `const` from the field.
`const` is hardly ever a good idea, I think. Usually you can achieve the same with data-hiding (i.e., making the field private and only provide a read-only accessor).

Data structures with private fields are needlessly cumbersome to use.

Reply:
Given how `abi::RegisterState::Map` is structured - it's either `const` or a template. And the second option leads to its own problems.

The point is that it only keeps track of relevant registers (limited by a single architecture) - const field is there to denote which architecture it is (it's passed in the constructor).

Reply:
Mmmh, why can't we simply drop `const`?

Reply:
Because it would make it easy to achieve invalid state. The const parameter determines both how many fields the container has and their relative position. By having a way to change it at runtime, a broken overall state becomes achievable. And for what? So that it can be assigned?

Reply:
How can the user lead to an invalid state?

> So that it can be assigned to?

Well... Yes! `:)` It cannot even be moved, it's a pretty rigid data structure.

Reply:
If we drop `const`, *a user of the class* (which is a programmer `:)`), how can it bring it into an invalid state without changing it?
If the answer is "it can't" I think there's no benefit and we should drop `const` in favor of making it assignable, which it can always come in handy.

Reply:
By "programmer" I meant "A person who might want to modify the class" over "a person who is going to use the class", (sorry for dodgy word selection 😉).

Technically, adding additional members (especially accessors) to a class feels less dangerous than removing a "const" from a field - so it's a lot more likely to happen.

Reply:
I think you're trying to defend against something you can't defend against. If I wanted to add a wrong accessor, I'd just drop `const` and go ahead and do it `:D`.

@drblallo suggests to add an assertion that `getArchitecture` doesn't return a non-`const` reference.
If you prefer we can store `FirstRegister` and `LastRegister`, but `const` fields are really annoying. In the model, we wanted to have them (for the key fields), but it made things so unergonomic that we decide to leave things public and postpone to when we had a generator to do data-hiding (which I really think is the proper way of handling the situation).

Reply:
(Now I'm starting to feel weird about thinking to bring up to you the fact that I wanted more `const` support from tuple tree generator for #161 😅)

but back on topic. Well yeah, I can't protect it from you, but it might just help someone who isn't as quick to drop all the `const`s they see :wink:.

Either way, ok, I give up. I'll send the patch to Antonio tomorrow.